### PR TITLE
 PLAT-641 | Stop using hardcoded payment schedule

### PR DESCRIPTION
**Pull request by rogatty**
https://bethink.atlassian.net/browse/PLAT-641

TODO:

- [x] Add tests

### CODE REVIEW

File: resources/assets/js/payment.js [hunk 1]
```diff
@@ -57,6 +58,11 @@ $(function () {
 		});
 	});
 
+	$('[data-instalment-due-date]').each(function () {
+		let date = moment.unix($(this).attr('data-timestamp'));
+		$(this).html(date.format('D MMMM'));
```
Comment (rogatty) [inline resources/assets/js/payment.js]: I wanted to format it server-side but Carbon requires `locales` to be installed (https://stackoverflow.com/a/32594258) and it's not a part of APK index that alpine uses and I couldn't rebuild the docker image:
```
Step 3/4 : RUN apk --no-cache add freetype-dev libjpeg-turbo-dev libpng-dev libzip-dev locales
 ---> Running in 05f17ea06d3d
fetch http://dl-cdn.alpinelinux.org/alpine/v3.9/main/x86_64/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.9/community/x86_64/APKINDEX.tar.gz
ERROR: unsatisfiable constraints:
  locales (missing):
    required by: world[locales]
ERROR: Service 'php' failed to build: The command '/bin/sh -c apk --no-cache add freetype-dev libjpeg-turbo-dev libpng-dev libzip-dev locales' returned a non-zero code: 1
```

File: app/Models/Order.php [hunk 2]
```diff
@@ -113,14 +113,19 @@ public function getInstalmentsAttribute()
...
-		// TODO: https://bethink.atlassian.net/browse/PLAT-641
-		$paymentDates = [
-			$this->created_at->addDays(7),
-			Carbon::createFromDate(2018, 11, 20),
-			Carbon::createFromDate(2018, 12, 20),
-		];
+		$orderInstalments = $this->orderInstalments()->get();
+
+		if ($orderInstalments->count() > 0) {
+			$orderInstalments = $this->generatePaymentSchedule();
```
Comment (kvas-damian) [inline app/Models/Order.php]: Why do we need regenerate instalments when they're already defined in DB?

File: resources/assets/js/payment.js (hunk 1 above)
Comment (kvas-damian) [reply resources/assets/js/payment.js]: you could try the solution provided in https://github.com/gliderlabs/docker-alpine/issues/144

File: app/Models/Order.php (hunk 2 above)
Comment (gitkuba) [reply app/Models/Order.php]: Yup, this line seems to be redundant.

File: app/Models/Order.php [hunk 3]
```diff
@@ -162,37 +167,53 @@ public function getInstalmentsAttribute()
 		];
 	}
 
+	/**
```
Comment (gitkuba) [inline app/Models/Order.php]: The above calculation should be entirely replaceable by the information from DB now.

Comment (gitkuba) [reply app/Models/Order.php]: Uhm, sorry. Couldn't get a right preview. I mean those lines: https://github.com/bethinkpl/wnl-platform/blob/7742092fc3f38890e8250add04a776653217fae9/app/Models/Order.php#L129-L160

File: app/Models/Order.php [hunk 4]
```diff
@@ -113,14 +113,19 @@ public function getInstalmentsAttribute()
...
 		$nextPayment = null;
-		$now = Carbon::now();
 
-		// TODO: https://bethink.atlassian.net/browse/PLAT-641
-		$paymentDates = [
-			$this->created_at->addDays(7),
-			Carbon::createFromDate(2018, 11, 20),
-			Carbon::createFromDate(2018, 12, 20),
-		];
+		$orderInstalments = $this->orderInstalments()->get();
```
Comment (bognix) [inline app/Models/Order.php]: why do you call `get` here, can't you just use `$this->orderInstalments`?

File: app/Models/Order.php [hunk 5]
```diff
@@ -162,37 +167,53 @@ public function getInstalmentsAttribute()
...
-			$valueToDistribute -= $amount;
-
-			if ($paidToDistribute >= $amount) {
-				$paidAmount = $amount;
-				$paidToDistribute -= $amount;
-			} else {
-				$paidAmount = $paidToDistribute;
-				$paidToDistribute = 0;
-			}
+		return $this->product->instalments()
```
Comment (bognix) [inline app/Models/Order.php]: how about adding unit tests to this function?

File: resources/assets/js/payment.js (hunk 1 above)
Comment (rogatty) [reply resources/assets/js/payment.js]: It's messy and I don't think is worth it. The current solution works just fine as these dates are hidden in a collapsed container anyway. I'd circle back to the topic if we ever want to optimize by removing moment.

File: app/Models/Order.php (hunk 4 above)
Comment (rogatty) [reply app/Models/Order.php]: No, because `$this->generatePaymentSchedule()` returns a `Collection` object and I want to keep the interface consistent.

Comment (bognix) [reply app/Models/Order.php]: what is returned then by `$this->orderInstalments` ?

File: app/Models/Order.php [hunk 6]
```diff
@@ -109,90 +110,79 @@ public function getCouponAmountAttribute()
...
-					'amount' => $instalment['left'],
-					'date'   => $date,
-				];
-			}
-			$totalLeft += $instalment['left'];
-		}
+		$orderInstalments = $this->generateAndSavePaymentSchedule();
+
+		$nextPayment = $orderInstalments
+			->sort(function (OrderInstalment $a, OrderInstalment $b) {
```
Comment (kvas-damian) [inline app/Models/Order.php]: Why do we need to sort them? Aren't they sorted when querying `ProductInstalments` model?

File: resources/views/payment/invoices/final.blade.php [hunk 7]
```diff
@@ -149,20 +150,21 @@
...
+						<td>{{ $n($instalment->amount) }}zł</td>
 					</tr>
 				@endif
 			@endforeach
 			<tr>
 				<td><strong>Razem</strong></td>
-				<td>{{ $n($invoiceOrder->instalments['total']/ 1.23) }}zł</td>
-				<td>{{ $n($invoiceOrder->instalments['total']- $invoiceOrder->instalments['total'] / 1.23) }}zł</td>
+				<td>{{ $n($invoiceOrder->instalments['total'] / 1.23) }}zł</td>
+				<td>{{ $n($invoiceOrder->instalments['total'] - $invoiceOrder->instalments['total'] / 1.23) }}zł</td>
```
Comment (kvas-damian) [inline resources/views/payment/invoices/final.blade.php]: For the future: we should avoid such calculations in view files.

File: app/Models/Order.php [hunk 8]
```diff
@@ -109,90 +110,79 @@ public function getCouponAmountAttribute()
...
+			'instalments' => $orderInstalments,
 			'nextPayment' => $nextPayment,
 			'total'       => $totalLeft,
 		];
 	}
 
+	/**
+	 * @return OrderInstalment[]|\Illuminate\Support\Collection
+	 */
 	public function generatePaymentSchedule()
```
Comment (kvas-damian) [inline app/Models/Order.php]: It's a nice candidate for a test. Especially with coupons applied.

**Pull request merged**